pkg/apis/eventing/v1alpha1/channel_defaults.go

```diff
@@ -16,11 +16,36 @@ limitations under the License.
 
 package v1alpha1
 
-//TODO replace this with openapi defaults when
-// https://github.com/kubernetes/features/issues/575 lands (scheduled for 1.13)
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// ChannelProvisionerDefaulter sets the default Provisioner and Arguments on Channels that do not
+// specify any Provisioner.
+type ChannelProvisionerDefaulter interface {
+	// GetChannel determines the default provisioner and arguments for the given channel. It does
+	// not modify the given channel. It may return nil for either or both.
```

Comment:
Seems if there's no objref, there's no point in returning the args? Or what does it mean if it returns objref but no args? Would be possibly better to just return nil for both if no objref.

Reply:
You're right, there is no point in setting the arguments if the provisioner is `nil`. But there is also no harm, is it worth calling out?

```golang
// GetDefault determines the default provisioner and arguments for the given channel. It does
// not modify the given channel. It may return nil for the provisioner. If the output of this function is
// used, then both the provisioner and arguments must be set (not just one or the other).
```

Reply:
Well, I'm just wondering about a mistake where configmap gets set so the arguments are there, but for some reason the provider is not. User creates something that uses default channel, it fails and they fill in a provider without arguments and the arguments come in from the defaulter, but the provisioner comes in from the user.

Reply:
In order to get any defaulting, the `provisioner` must be `nil`. So it can't be that the user provides the `provisioner` and the defaulter provides the `arguments`.  

If the user initially tries to get a default `provisioner` (by leaving it `nil` in the Channel's Spec). If the `provisioner` is not filled in by the defaulter, then the webhook will reject the `Channel`. If for some reason it got in and the user then changed the `provisioner`, then the webhook would reject the update (`provisioner` is an immutable field).

Reply:
Ping @vaikas-google Is my explanation sufficient? Or would you like more explanation and/or code changes?